### PR TITLE
PR for #180: More informative error messages when list is used as a key

### LINKED ISSUE #180: SaveData sends uninformative error messages when key column contains lists
When specifying a key column that contains list elements, SaveData throws out uninformative error:
```
Traceback (most recent call last):
  File "/Users/zpu/Documents/GitHub/WorldOrder/source/derived/ISO/iso.py", line 140, in <module>
    Main()
  File "/Users/zpu/Documents/GitHub/WorldOrder/source/derived/ISO/iso.py", line 48, in Main
    SaveData(transitions, 
    ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/zpu/Documents/GitHub/WorldOrder/source/lib/SaveData.py", line 14, in SaveData
    CheckKeys(df, keys)
  File "/Users/zpu/Documents/GitHub/WorldOrder/source/lib/SaveData.py", line 43, in CheckKeys
    if not all(df.groupby(keys).size() == 1):
               ^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/anaconda3/envs/pt/lib/python3.11/site-packages/pandas/core/groupby/groupby.py", line 3043, in size
    result = self._grouper.size()
             ^^^^^^^^^^^^^^^^^^^^
  File "/opt/anaconda3/envs/pt/lib/python3.11/site-packages/pandas/core/groupby/ops.py", line 705, in size
    ids, _, ngroups = self.group_info
                      ^^^^^^^^^^^^^^^
  File "properties.pyx", line 36, in pandas._libs.properties.CachedProperty.__get__
  File "/opt/anaconda3/envs/pt/lib/python3.11/site-packages/pandas/core/groupby/ops.py", line 745, in group_info
    comp_ids, obs_group_ids = self._get_compressed_codes()
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/anaconda3/envs/pt/lib/python3.11/site-packages/pandas/core/groupby/ops.py", line 764, in _get_compressed_codes
    group_index = get_group_index(self.codes, self.shape, sort=True, xnull=True)
                                  ^^^^^^^^^^
  File "/opt/anaconda3/envs/pt/lib/python3.11/site-packages/pandas/core/groupby/ops.py", line 690, in codes
    return [ping.codes for ping in self.groupings]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/anaconda3/envs/pt/lib/python3.11/site-packages/pandas/core/groupby/ops.py", line 690, in <listcomp>
    return [ping.codes for ping in self.groupings]
            ^^^^^^^^^^
  File "/opt/anaconda3/envs/pt/lib/python3.11/site-packages/pandas/core/groupby/grouper.py", line 691, in codes
    return self._codes_and_uniques[0]
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "properties.pyx", line 36, in pandas._libs.properties.CachedProperty.__get__
  File "/opt/anaconda3/envs/pt/lib/python3.11/site-packages/pandas/core/groupby/grouper.py", line 835, in _codes_and_uniques
    codes, uniques = algorithms.factorize(  # type: ignore[assignment]
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/anaconda3/envs/pt/lib/python3.11/site-packages/pandas/core/algorithms.py", line 795, in factorize
    codes, uniques = factorize_array(
                     ^^^^^^^^^^^^^^^^
  File "/opt/anaconda3/envs/pt/lib/python3.11/site-packages/pandas/core/algorithms.py", line 595, in factorize_array
    uniques, codes = table.factorize(
                     ^^^^^^^^^^^^^^^^
  File "pandas/_libs/hashtable_class_helper.pxi", line 7281, in pandas._libs.hashtable.PyObjectHashTable.factorize
  File "pandas/_libs/hashtable_class_helper.pxi", line 7195, in pandas._libs.hashtable.PyObjectHashTable._unique
TypeError: unhashable type: 'list'
```

I see 2 options to fix this:
1. Explicitly ban lists from being used as keys; generate an error: lists detected in one of the keys
2. Allow keys columns to contain list elements.

### PR DESCRIPTION
**Pull request by liaochris**
In this issue, I updated SaveData.py to provide more informative error messages when users provide keys that are of type list. I also added a test case that ensures a type error is raised.

### CODE REVIEW

**Comment by liaochris** [comment]
Will be done here @zhizhongpu

**Comment by liaochris** [comment]
Updated from master in https://github.com/gslab-econ/gslab_python/commit/36be2c41c60f539316ea1dc7bafe13fa34985605

**Review by zhizhongpu — COMMENTED** [review]
@liaochris this looks great - left a few comments.

File: gslab_misc/SaveData/tests/test_main.py [hunk 1]
```diff
@@ -20,7 +21,20 @@ def test_wrong_keytype(self):
...
+        df = pd.read_csv('data/data.csv')
+        df['bad_id'] = df['id'].apply(lambda x: [x])
+        with self.assertRaises(TypeError):
+            SaveData(df, 'bad_id', 'dfs.csv')        
+
+    def test_wrong_keytype_list_one(self):
+        df = pd.read_csv('data/data.csv')
+        df['bad_id'] = df['id'].apply(lambda x: [x] if x == 1 else x)
+        with self.assertRaises(TypeError):
+            SaveData(df, 'bad_id', 'dfs.csv')
```
[the remark this answers is not in the document]
Comment (liaochris) [reply gslab_misc/SaveData/tests/test_main.py:35]: @zhizhongpu the second test is to ensure that the export fails when we have columns of mixed type (ie: only one row is a list). 

Also I agree that we should add an additional test to make sure that we cannot accept any columns of type list when exporting data.

[the remark this answers is not in the document]
Comment (liaochris) [reply gslab_misc/SaveData/tests/test_main.py]: Added new test case for column of type list (not explicitly key) + renamed `bad_id` to `list_id`

**Review by zhizhongpu — APPROVED** [review]
@liaochris the latest version passed the tests so approving. Note that I've done a small tweak in d5a03def1c43355022e00e3bdfb4d6e844609151. If it looks good feel free to merge; if not let's discuss. Thanks

**Comment by liaochris** [comment]
Summary comment https://github.com/gslab-econ/gslab_python/issues/180#issuecomment-2349003495

**Pull request merged**